ARROW-9523 [Rust] Improve filter kernel performance

yordan-pavlov opened the pull request:
The filter kernel located here https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/kernels/filter.rs

currently has the following performance:

filter old u8 low selectivity time: [1.7782 ms 1.7790 ms 1.7801 ms]
filter old u8 high selectivity time: [815.58 us 816.58 us 817.57 us]
filter old u8 w NULLs low selectivity time: [1.8131 ms 1.8231 ms 1.8336 ms]
filter old u8 w NULLs high selectivity time: [817.41 us 820.01 us 823.05 us]

I have been working on a new implementation which performs between approximately 17 and 550 times faster depending mostly on filter selectivity. Here are the benchmark results:

filter u8 low selectivity                  time:   [107.26 us 108.24 us 109.58 us]
filter u8 high selectivity                 time:   [4.7854 us 4.8050 us 4.8276 us]
filter context u8 low selectivity          time:   [102.59 us 102.93 us 103.38 us]
filter context u8 high selectivity         time:   [1.4709 us 1.4760 us 1.4823 us]
filter context u8 w NULLs low selectivity  time:   [130.48 us 131.00 us 131.65 us]
filter context u8 w NULLs high selectivity time:   [2.0520 us 2.0818 us 2.1137 us]
filter context f32 low selectivity         time:   [117.26 us 118.58 us 120.13 us]
filter context f32 high selectivity        time:   [1.7895 us 1.7919 us 1.7942 us]

This new implementation is based on a few key ideas:

(1) if the data array being filtered doesn't have a null bitmap, no time should be wasted to copy or create a null bitmap in the resulting filtered data array - this is implemented using the CopyNullBit trait which has a no-op implementation and an actual implementation

(2) when the filter is highly selective, e.g. only a small number of values from the data array are selected, the filter implementation should efficiently skip entire batches of 0s in the filter array - this is implemented by transmuting the filter array to u64 which allows to quickly check and skip entire batches of 64 bits 

(3) when an entire record batch is filtered, any computation which only depends on the filter array is done once and then shared for filtering all the data arrays in the record batch - this is implemented using the FilterContext struct

This pull request also implements support for filtering dictionary arrays. 

@paddyhoran @andygrove

wesm commented:
I refer you all to the work that we've recently done in C++ where we use popcount on 64 bits at a time to efficiently compute the size of the filter output as well as quickly compute the filtered output. It might be worth replicating the "BitBlockCounter" concept in Rust

andygrove commented:
@yordan-pavlov This is exciting! I will start reviewing this later today.

yordan-pavlov commented:
just for reference, I think the relevant C++ filter implementation (which wesm is referring to) is in the PrimitiveFilterImpl class here https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/vector_selection.cc#L558

yordan-pavlov commented on rust/arrow/src/compute/kernels/filter.rs:
```diff
@@ -17,139 +17,443 @@
+    let filtered_arrays = record_batch
+        .columns()
+        .iter()
+        .map(|a| filter_context.filter(a.as_ref()).unwrap())
```
good spot, I have changed the filter_record_batch method to remove the unwrap() and use `.collect::<Result<Vec<ArrayRef>>>()?;` instead

yordan-pavlov commented:
@jorgecarleitao thanks for the feedback

I did some more profiling specifically around the unsafe parts of code and found that the safe version of `copy_null_bit` is just as fast so have removed that unsafe section; here are some benchmark results:

copy_null_bit unsafe:
filter context u8 w NULLs low selectivity  time:   [142.05 us 142.35 us 142.68 us]
filter context u8 w NULLs high selectivity time:   [2.0915 us 2.1015 us 2.1127 us]

copy_null_bit safe:
filter context u8 w NULLs low selectivity  time:   [134.74 us 134.86 us 134.98 us]
filter context u8 w NULLs high selectivity time:   [2.0536 us 2.0613 us 2.0707 us]

I also benchmarked replacing the unsafe section in the `filter_array_impl` method with `value_buffer.write()` but this results in approximately 17% drop in performance with sparse filter arrays as can be seen from the benchmark results below:

filter u8 low selectivity
                        time:   [131.08 us 132.46 us 134.27 us]
                        change: [+13.141% +17.189% +22.115%] (p = 0.00 < 0.05)

filter context u8 low selectivity
                        time:   [127.47 us 129.27 us 131.56 us]
                        change: [+12.008% +19.674% +27.939%] (p = 0.00 < 0.05)

filter context u8 w NULLs low selectivity
                        time:   [154.32 us 155.27 us 156.79 us]
                        change: [+15.444% +17.846% +22.268%] (p = 0.00 < 0.05)

filter context f32 low selectivity
                        time:   [137.62 us 138.01 us 138.52 us]
                        change: [+12.495% +18.180% +23.088%] (p = 0.00 < 0.05)

finally, looking at the C++ implementation inspired me to change the `filter_array_impl` method to add a special case where the 64bit filter batch is all 1s and this doesn't appear to reduce performance in other cases but improves performance of filtering with very dense filter arrays (almost all 1s) by about 20 times; here are the latest benchmark results:

filter u8 low selectivity                      time:   [109.75 us 110.30 us 111.02 us]
filter u8 high selectivity                     time:   [4.8372 us 4.8433 us 4.8502 us]
filter u8 very low selectivity                 time:   [11.782 us 11.798 us 11.816 us]
filter context u8 low selectivity              time:   [109.07 us 109.65 us 110.66 us]
filter context u8 high selectivity             time:   [1.4704 us 1.4762 us 1.4842 us]
filter context u8 very low selectivity         time:   [8.8455 us 9.0530 us 9.3171 us]
filter context u8 w NULLs low selectivity      time:   [135.32 us 135.49 us 135.66 us]
filter context u8 w NULLs high selectivity     time:   [2.0579 us 2.0680 us 2.0796 us]
filter context u8 w NULLs very low selectivity time:   [11.583 us 11.668 us 11.780 us]
filter context f32 low selectivity             time:   [138.71 us 139.83 us 141.41 us]
filter context f32 high selectivity            time:   [1.6111 us 1.6342 us 1.6605 us]
filter context f32 very low selectivity        time:   [19.719 us 19.865 us 20.045 us]

@andygrove  any thoughts?

andygrove commented:
Sorry @yordan-pavlov but I didn't get a chance to review yet. I hope to get to it soon.

andygrove commented on rust/arrow/src/compute/kernels/filter.rs:
```diff
@@ -17,139 +17,466 @@
+            let data_index = i * 64;
+            let data_len = value_size * 64;
+            null_bit_setter.copy_null_bits(data_index, 64);
+            unsafe {
```
The use of `unsafe` makes me nervous. If the performance improvement justifies using this, how can we be sure we have adequate testing for this?

yordan-pavlov replied:
@andygrove  I have replaced the use of unsafe memcpy with copy_from_slice - this does reduce performance (by about 12%) but it is still much faster than the old implementation; here are the latest benchmark results:

filter u8 low selectivity                      time:   [130.74 us 131.37 us 132.19 us]
filter u8 high selectivity                     time:   [5.2031 us 5.2366 us 5.2764 us]
filter u8 very low selectivity                 time:   [12.353 us 12.542 us 12.759 us]

filter context u8 low selectivity              time:   [129.54 us 129.88 us 130.30 us]
filter context u8 high selectivity             time:   [1.7926 us 1.7974 us 1.8046 us]
filter context u8 very low selectivity         time:   [8.7700 us 8.7987 us 8.8342 us]

filter context u8 w NULLs low selectivity      time:   [150.36 us 151.09 us 152.01 us]
filter context u8 w NULLs high selectivity     time:   [2.4173 us 2.4882 us 2.5703 us]
filter context u8 w NULLs very low selectivity time:   [158.86 us 160.32 us 162.26 us]

filter context f32 low selectivity             time:   [123.38 us 124.34 us 126.18 us]
filter context f32 high selectivity            time:   [1.4836 us 1.4994 us 1.5297 us]
filter context f32 very low selectivity        time:   [19.422 us 19.653 us 19.932 us]

andygrove replied:
Thanks @yordan-pavlov We can always consider the unsafe changes in a future PR, but removing them for now makes it much easier to get these changes merged.

andygrove commented on rust/arrow/src/compute/kernels/filter.rs:
```diff
@@ -17,139 +17,466 @@
+                let data_index = (i * 64) + j;
+                null_bit_setter.copy_null_bit(data_index);
+                // if filter bit == 1: copy data value to temp array
+                unsafe {
```
Same comment as above re `unsafe`

andygrove commented:
LGTM overall.

yordan-pavlov commented:
@paddyhoran yes, you are right, I added a couple more tests for sliced arrays and they didn't pass so seeing that the PR was not yet merged I added a few small changes to 
(1) implement support for filtering of sliced / offset data arrays
(2) return an error if the filter array is offset - I thought it better to make this obvious rather than hide it

I looked briefly into implementing support for offset filter arrays but I couldn't figure out how to do it quickly - I might have to have another look at the C++ implementation and submit a separate PR for that.

wesm commented:
This PR makes me think that at some point (when someone gets really motivated), it would be interesting to implement a non-language-dependent benchmark harness for certain operations (such that can be represented using e.g. a standard protobuf serialized operation/expression) so that we can get apples-to-apples numbers for certain operations across implementations. It would be interesting to know e.g. how the Rust implementation of filtering compares with the C++

andygrove commented:
I completely agree. I just happen to have a protobuf definition for plans and expressions as well as serde code for Java and Rust that I would be happy to contribute. We would need to implement for C++.
